django/contrib/auth/hashers.py

```diff
@@ -515,3 +515,67 @@ def safe_summary(self, encoded):
             (_('salt'), salt),
             (_('hash'), mask_hash(data, show=3)),
         ])
+
+
+class Argon2PasswordHasher(BasePasswordHasher):
+    """
+    Secure password hashing using the argon2 algorithm
+
+    This is the winner of the Password Hashing Competition
+    ( https://password-hashing.net ).  You must first install the
+    argon2-cffi library.  Please be warned that this library depends
+    on native C code and might cause portability issues.
+    """
+    algorithm = "argon2"
+    library = "argon2"
+
+    time_cost = 2
+    memory_cost = 512
+    parallelism = 2
+
+    def encode(self, password, salt):
+        argon2 = self._load_library()
+        data = argon2.low_level.hash_secret(force_bytes(password),
+                                            force_bytes(salt),
+                                            time_cost=self.time_cost,
+                                            memory_cost=self.memory_cost,
+                                            parallelism=self.parallelism,
+                                            hash_len=16,
```

Comment:
Done.  We do have to be careful: if `argon2.DEFAULT_HASH_LENGTH` is increased significantly in the future, `encode` can return a string longer than the maximum 128.